### ISSUE TITLE
Support layout `responsive` for `carousel type=carousel`

### ISSUE THREAD

**Issue by aghassemi** [ampproject/amphtml#5977]
I am not clear why layout `responsive` is not supported for `carousel` same way it is for `slides`. Seems like something we should be able to support.

/cc @dvoytenko @muxin @camelburrito

**Comment by dvoytenko** [comment]
The problem there is that we can't do it via CSS. With our CSS hacks we can only condition height to be a ratio of width, not the other way around. This works well with slides because width is known - it's equal to container width. However, with scrollable carousel, we can't rely on anything like this for width - width is variable. Instead, we want to fix height and have width follow the aspect ratio. And this is not something we can do unfortunately.

**Comment by aghassemi** [comment]
@dvoytenko I don't fully understand. Considering `layout=responsive` applies to the carousel itself not the items in it, I don't see how it is different than slides. Any layout that is `isLayoutSizeDefined` would end up creating a presized box, all carousel needs to do is to have the `<div>` that contains the items to fill the parent `<anp-carousel>` similar to how slides does it and we should be good, No? I tried this approach and seems to work for basic cases and I can't find a situation where it would fail.

**Comment by aghassemi** [comment]
@dvoytenko @camelburrito @muxin I am suspicious not allowing `layout=responsive` on carousel might have been a misunderstanding. I understands we can't support layout responsive for `items` inside the carousel the same way we can for `slides`, but that has nothing to do with `layout=responsive` on carousel itself.

**Comment by camelburrito** [comment]
layout=responsive on the carousel - kills the fixed height aspect of carousel  - which is required to guarantee layout="responsive" on the inner items - (because the responsive resize happens with width changing and the runtime/css computing the height - and does not work the other way).

@dvoytenko is this what you meant?

**Comment by aghassemi** [comment]
> which is required to guarantee layout="responsive" on the inner items
`layout=responsive` on inner items does not currently work at all. 

Responsive width for items inside carousel is somewhat meaningless but we can actually support it if we wanted to by giving it some documented meaning, such as width of responsive inner items will be 100% of carousel width (which would make carousel similar to slides without the snapping) or we could even say width of responsive inner items will be 60% of carousel width allowing 20% padding on each side so next/previous items are also visible.

**Comment by camelburrito** [comment]
It won't be 100% because there is a set height on the parent , width varies with respect to height.

**Comment by aghassemi** [comment]
Sure, it may makes more sense to want to vary width based on height for responsive inner items in the case of carousel (different than what responsive does in other places however) but I think this is Dima's point that our CSS tricks can only vary height with respect to width and so it will be hard to support this.

Regardless, supporting or not supporting responsive items inside carousel is fine but has nothing to do with actually supporting layout responsive on the carousel itself, which is the main point of this issue.

**Comment by dvoytenko** [comment]
It was mostly expectation. Because on the surface, this sounds easy enough: we have all the necessary data to calculate the right heights and widths correctly. It's just that it's really hard to do in runtime. So we wanted to avoid the expectation that an author sets `layout=responsive` on the carousel, tests on one screen size, deploys and then it turns out we can't correctly resize children.

That being said - we can't correctly resize via CSS. It might be worth considering to do it in JS. We incur some flash, but it might end up being minimal given that carousel already controls the visibility of its children. Might be worth trying.

**Comment by aghassemi** [comment]
@dvoytenko Isn't layout `responsive` just `fixed-height width=auto` where height is calculated from `width/height` ratio (and recalculated when width changes)? As far as `amp-carousel` is concerned, the height is already fixed by runtime regardless of `fixed-height` or `responsive` so I still fail to see why carousel can't support `responsive` same way it supports `fixed-height`. Exactly where in the code does it care what the height of carousel is? I don't see any height measurement in scrollable-carousel. I really don't see the difference between `slides` and `carousel` when it comes to `layout` of the parent (again I understand `responsive` layout for `children` is different between them but that's about it. scrollable-carousel doesn't currently support responsive children anyway)

(The reason I am bringing this issue up is that there are a lot of user errors out there setting `layout responsive` for `type=carousel` since it is not something validator catches. So this is not some low priority feature request)

**Comment by aghassemi** [comment]
Summarizing the quick chat we had with @cramforce and @camelburrito:

We should:
1- Allow responsive children on carousel, but change the meaning of responsive from width to height. (e.g. Css will do the opposite of what current responsive css does, it sets the height 100% and padding left to the ratio.) We may need to add a new layout=`height-responsive` if it is too late for carousel to change the meaning of layout for its children at its build time.
2- Allow any layout on carousel itself.
3- Document very well what layouts are supported for children. Looks like we can support any layout for children as well ( we need to make sure `fill` behaves correctly as well, filling only the visible width/height essentially making carousel into slides without snap points )

**Comment by dvoytenko** [comment]
`padding-left` would still be width-based. So, this probably won't give us aspect ratio. Or am I missing something?

**Comment by dvoytenko** [comment]
Ok, so here's what I think we should do:

1. Indeed allow `layout=responsive` on scrollable carousel.
2. Allow children of scrollable carousel to have layout parameters (see https://github.com/ampproject/amphtml/blob/master/spec/amp-html-layout.md). This would mainly include `width`, `height` and `layout`. This includes both: AMP and non-AMP elements.
3. Create a utility that will create a JS-based implementation of height-based aspect-ratio `layout=responsive`.
4. Use this utility in scrollable carousel via `onLayoutMeasure`.

**Comment by techjoshua** [comment]
I'm glad to see you guys are working to support `type="carousel" layout="responsive"`.
I've been trying to figure out how I can make a carousel full of square child elements to be as wide as the viewport, and have its height adjust so I can choose to fully show 2, 3, or 4 carousel items based on the viewport dimensions.

I've hacked together something that works, but of course it doesn't validate.  I've had to style the `amp-carousel` element and its children with media queries and drop the width, height, and layout attributes which the validator complains about.

The funny thing is that it works... even though the validator isn't happy I can get the page to work as desired.  And in theory it doesn't violate the principles on which AMP is based, as the dimensions are deterministic.  My CSS:
```
amp-carousel {
	height: 50vw;
	width: 100vw;
	position: relative;
}
			
amp-carousel .slide {
	display: inline-block;
	width: 50vw;
	height: 100%;
}
			
@media (min-width: 767px) {
	amp-carousel {
		height: 33.333vw;
	}
	amp-carousel .slide {
		width: 33.333vw;
		height: 33.333vw;
	}
}
			
@media (min-width: 991px) {
	amp-carousel {
		height: 25vw;
	}
	amp-carousel .slide {
		width: 25vw;
		height: 25vw;
	}
}
```

**Comment by dvoytenko** [comment]
I believe @aghassemi found an auto-resizing solution via shape-defining SVG.

**Comment by aghassemi** [comment]
@techjoshua The solution we like to implement here is a responsive behaviour based on height and auto-resizing the width based on height of each item being 100% of the carousel without distorting the aspect ratio. That would allow items with different aspect ratios to mix with each other and removes the need to do breakpoints.

**Comment by aghassemi** [comment]
let's evaluate in the context of `amp-stream-gallery`

**Comment by nainar** [comment]
This issue is fixed for all users who upgrade their `<amp-carousel>` version to 0.2. 

For more details please see documentation here: https://github.com/ampproject/amphtml/blob/master/extensions/amp-carousel/amp-carousel.md
